### PR TITLE
LRU with ring buffer

### PR DESCRIPTION
**Pull request by ldorau**
This change is [Reviewable](https://reviewable.io/reviews/pmem/vmemcache/139)

### CODE REVIEW

File: src/vmemcache_repl.c [hunk 1]
```diff
@@ -196,17 +200,45 @@ repl_p_lru_new(struct repl_p_head **head)
...
 	TAILQ_INIT(&h->first);
+	h->ringbuf = ringbuf_new(LEN_RING_BUF);
 	*head = h;
 
 	return 0;
 }
 
+/*
+ * dequeue_all -- (internal) dequeue all repl_p entries,
+ *                           it MUST be run under WR lock
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/vmemcache_repl.c]: Done

File: src/vmemcache_repl.c [hunk 2]
```diff
@@ -196,17 +200,45 @@ repl_p_lru_new(struct repl_p_head **head)
...
+						sizeof(struct repl_p_entry));
+		if (e == NULL)
+			break;
+
+		TAILQ_MOVE_TO_TAIL(&head->first, e, node);
+		util_bool_compare_and_swap64(e->ptr_entry, NULL, e);
+
+		counter++;
+
+	} while (counter < LEN_RING_BUF);
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/vmemcache_repl.c]: Done

File: src/vmemcache_repl.c [hunk 3]
```diff
@@ -250,14 +282,24 @@ repl_p_lru_insert(struct repl_p_head *head, void *element,
...
 	ASSERTne(ptr_entry, NULL);
 
-	util_mutex_lock(&head->lock);
+	entry = util_val_compare_and_swap64(ptr_entry, *ptr_entry, NULL);
+	if (entry == NULL)
+		return;
 
-	if (*ptr_entry != NULL)
-		TAILQ_MOVE_TO_TAIL(&head->first, *ptr_entry, node);
+	if (ringbuf_tryenqueue(head->ringbuf, entry) == 0)
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/vmemcache_repl.c]: Done

File: src/vmemcache_repl.c [hunk 4]
```diff
@@ -267,30 +309,37 @@ static void *
...
-	TAILQ_REMOVE(&head->first, entry, node);
+	if (ptr_entry == NULL)
+		ptr_entry = TAILQ_FIRST(&head->first)->ptr_entry;
 
-	ASSERTne(entry->ptr_entry, NULL);
-	*(entry->ptr_entry) = NULL;
+	int max_retry = 2;
 
-	util_mutex_unlock(&head->lock);
+	do {
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/vmemcache_repl.c]: Done

**Comment by ldorau** [comment]
@kilobyte please review

**Comment by ldorau** [comment]
Rebased

**Comment by ldorau** [comment]
Rebased on top of #138

**Comment by ldorau** [comment]
Rebased on master

File: src/vmemcache_repl.c [hunk 5]
```diff
@@ -196,17 +200,50 @@ repl_p_lru_new(struct repl_p_head **head)
...
+	int counter = 0;
+
+	do {
+		e = ringbuf_trydequeue_s(head->ringbuf,
+						sizeof(struct repl_p_entry));
+		if (e == NULL)
+			break;
+
+		TAILQ_MOVE_TO_TAIL(&head->first, e, node);
+		util_bool_compare_and_swap64(e->ptr_entry, NULL, e);
```
Comment (marcinslusarz) [inline src/vmemcache_repl.c]: Why do you even use compare&swap if you are not interested in compare result?

File: src/vmemcache_repl.c [hunk 6]
```diff
@@ -232,7 +269,7 @@ repl_p_lru_insert(struct repl_p_head *head, void *element,
 
 	ASSERTne(ptr_entry, NULL);
 	entry->ptr_entry = ptr_entry;
-	*(entry->ptr_entry) = entry;
+	util_bool_compare_and_swap64(entry->ptr_entry, NULL, entry);
```
Comment (marcinslusarz) [inline src/vmemcache_repl.c]: ditto

File: src/vmemcache_repl.c [hunk 7]
```diff
@@ -250,14 +287,19 @@ repl_p_lru_insert(struct repl_p_head *head, void *element,
...
 {
-	ASSERTne(ptr_entry, NULL);
+	struct repl_p_entry *entry;
 
-	util_mutex_lock(&head->lock);
+	ASSERTne(ptr_entry, NULL);
 
-	if (*ptr_entry != NULL)
-		TAILQ_MOVE_TO_TAIL(&head->first, *ptr_entry, node);
+	entry = util_val_compare_and_swap64(ptr_entry, *ptr_entry, NULL);
```
Comment (marcinslusarz) [inline src/vmemcache_repl.c]: What's the point of compare&swap if you don't care about the previous value?

File: src/vmemcache_repl.c [hunk 8]
```diff
@@ -267,30 +309,36 @@ static void *
...
-		entry = TAILQ_FIRST(&head->first);
+	if (TAILQ_EMPTY(&head->first)) {
+		errno = ESRCH;
+		goto exit_unlock;
+	}
+
+	if (ptr_entry == NULL)
+		ptr_entry = TAILQ_FIRST(&head->first)->ptr_entry;
+
+	int max_retry = 2;
```
Comment (marcinslusarz) [inline src/vmemcache_repl.c]: ???

File: src/vmemcache_repl.c [hunk 9]
```diff
@@ -267,30 +309,36 @@ static void *
...
+
+	while ((entry = util_val_compare_and_swap64(ptr_entry,
+			*ptr_entry, NULL)) == NULL && (--max_retry > 0)) {
```
Comment (marcinslusarz) [inline src/vmemcache_repl.c]: same question as in repl_p_lru_use

File: src/vmemcache_repl.c (hunk 8 above)
Comment (ldorau) [reply src/vmemcache_repl.c]: What does '???' mean?

File: src/vmemcache_repl.c (hunk 7 above)
Comment (ldorau) [reply src/vmemcache_repl.c]: I do care. 'entry' is the previous value.

File: src/vmemcache_repl.c (hunk 5 above)
Comment (ldorau) [reply src/vmemcache_repl.c]: It has to be true. I can add assert() here.

File: src/vmemcache_repl.c (hunk 6 above)
Comment (ldorau) [reply src/vmemcache_repl.c]: It has to be true. I can add assert() here.

File: src/vmemcache_repl.c (hunk 9 above)
Comment (ldorau) [reply src/vmemcache_repl.c]: I do care. 'entry' is the previous value.

**Comment by ldorau** [comment]
I have just pushed the '__sync_bool_compare_and_swap' version.

**Pull request merged**